### ISSUE TITLE
Project whose path contains non-ASCII characters fails to build on Windows

### ISSUE THREAD

**Issue by mrbraitwistle** [commercialhaskell/stack#2810]
### Steps to reproduce

A project produced with these steps built without any errors on Ubuntu 14.04 but failed on Windows 10. 

1. Create a new project using `stack new test-project` in a directory whose name contains non-ASCII characters, such as `C:\Desktop\folderß\`. The project name itself doesn't have to contain non-ASCII characters.
2. Run command `stack build` inside `C:\Desktop\folderß\test-project\`.

### Expected

I expected stack to successfully build the project and not care about what directory it's in.

### Actual

The project was not successfully built, no binaries were produced.

```
$ stack build --verbose
```
[Output of `stack build --verbose` when building a project in a directory whose name contains non-ASCII characters on Windows](https://gist.github.com/mrbraitwistle/6eafa68030ac9ced0b53decf7bfe7e28)

### Stack version

```
$ stack --version
Version 1.2.0 x86_64 hpack-0.14.0
```

### Method of installation

Official binary, downloaded from https://docs.haskellstack.org/en/stable/install_and_upgrade, then upgraded using `stack upgrade`

**Comment by mgsloan** [comment]
Relevant portion of the log:

```
2016-11-26 21:19:23.190854: [info] Linking .stack-work\dist\2672c1f3\build\test-project-exe\test-project-exe.exe ...
2016-11-26 21:19:23.694368: [warn] C:/Users/mrbraitwistle/AppData/Local/Programs/stack/x86_64-windows/ghc-7.10.3/mingw/bin/ld.exe: cannot find -lHStest-project-0.1.0.0-8YNZKLV16RsHF0M0p5VrYR
@(Stack\Build\Execute.hs:920:67)
2016-11-26 21:19:23.695868: [warn] collect2.exe: error: ld returned 1 exit status
```

This indicates that it is an upstream issue with cabal.  Does `stack setup --upgrade-cabal` resolve the issue?

**Comment by mrbraitwistle** [comment]
No,  the problem persists after running `stack setup --upgrade-cabal`. The log looks almost the same, the only difference I am able to find is the following additional output after the relevant portion of the log:

```
2016-11-27 10:33:44.268764: [warn] collect2.exe: error: ld returned 1 exit status
@(Stack\Build\Execute.hs:920:67)
2016-11-27 10:33:44.273784: [warn] `gcc.exe' failed in phase `Linker'. (Exit code: 1)
@(Stack\Build\Execute.hs:920:67)
```

The only other difference is that the message about `haskeline` is not in the new log and the Cabal version in the message about Cabal is now different.

```
2016-11-26 21:19:22.482853: [debug] Ignoring package haskeline due to wanting version 0.7.2.3 instead of 0.7.2.1
@(Stack\Build\Installed.hs:196:5)
2016-11-26 21:19:22.483353: [debug] Ignoring package Cabal due to wanting version 1.22.8.0 instead of 1.22.5.0
@(Stack\Build\Installed.hs:196:5)
```
